// matcha/lib/matcha/bench.js
/*!
 * Matcha - Benchmark Constructor
 * Copyright(c) 2011-2012 Jake Luer <jake@alogicalparadox.com>
 * MIT Licensed
 */

/*!
 * External Dependancies
 */

var EventEmitter = require('events').EventEmitter
  , util = require('util');

/*!
 * Internal Dependancies
 */

var Timer = require('./timer');

/*!
 * Primary Exports
 */

module.exports = Bench;

function Bench (title, fn) {
  EventEmitter.call(this);
  this.title = title;
  this.fn = fn;
}

/*!
 * Inherits from `EventEmitter`
 */

util.inherits(Bench, EventEmitter);

/**
 * .run (cb)
 *
 * Run a bench.
 *
 * @param {Function} callback
 */

Bench.prototype.run = function (cb) {
  var self = this
    , opts = this.parent.options
    , runner = opts.type === 'static'
      ? runStatic
      : runAdaptive

  cb = cb || function () {};

  function report (total, elapsed) {
    var stats = {
        iterations: total
      , elapsed: elapsed
      , title: self.title
      , ops: Math.round(total / elapsed * 1000)
    };
    self.stats = stats;
    self.emit('end', stats);
    cb(stats);
  }

  runner.call(this, report);
};

/*!
 * runStatic (timer, cb)
 *
 * Run this bench as  static (set iterations).
 *
 * @param {Function} callback
 * @api private
 */

function runStatic (cb) {
  var self = this
    , parent = this.parent
    , iterations = parent.options.iterations
    , concurrency = parent.options.concurrency
    , pending = iterations
    , timer = new Timer()
    , total = 0;

  var runner = this.fn.length === 0
    ? doSync
    : doAsync;

  function
