// assertions/higher.js
var extended = require('./extended')
  , elementary = require('./elementary')
  , traverser = require('traverser')
  , render = require('render')
  , fail = require('./failure').fail

module.exports = {
  every:    every,
  has:      has,
  apply:    apply,
  throws:   throws,
  path:     path,
  noop:     noop,
  atIndex:  atIndex,
  property: property,
  hasKeys:  hasKeys
}
function property (actual, property, value, message){

    function explain(err, template) {
      throw fail(err).explain(
                'property:'+template
              , { actual: actual
                , property: property
                , path: [property]
                , assertion: value
                }
              , message 
              )     
    }

    if(!actual[property] && value == null)
    //checks that property is defined on actual, even if it is undefined (but not deleted)
      explain(new Error(), 'property: {actual:render}{path:path} must exist')
    //if value is a function, assume it is an assertion... ap
