// macgyver/rules.js
var assert = require("assert")

function ifNull(n) {
  return n == null ? -1 : n
}
function p(s) {
  return '('+s+')'
}
function bt(s) {
  return '`' + s + '`'
}

function abbrev (contract) {

  var func = contract.function.toString()
  var name = contract.name

  return [
    (name ? bt(name) : ''),
    p( func.length < 80
      ? func : func.substring(0, 100) + '...')
  ].join(': ')
}

function between (called, min, max) {
  if(max != null && called > max) return false
  if(min != null && called < min) return false
  return true
}

function plural (n) {
  return n > 1 || n === 0 ? ' times' : ' time'
}

function rangeDesc(min, max) {
  min = min == null ? null : min
  max = max == null ? null : max

  var s = ({
    '1:1'   : '*must* be called exactly once',
    '0:0'   : '*must* not be called',
    'null:1': '*may* be called',
    '1:null': '*must* eventually be called'
  })[min+':'+max]

  if (s) return s

  var a = (
      min == null ? '' :
      'at least ' + min + plural(min)
    )
  var b = (
       max == null ? '' :
      'at most ' + max + plural(max)
    )
  s = '*must* be called ' + (a && b ? a + ', and ' + b : a || b)

    return s
}

exports.isCalled = function (min, max) {
    if(min != null && max != null)
      if (min > max) throw new Error('min must be smaller than max')
   var err = new Error()
   function error (con) {
      err.message = (
        'broke contract: '
        + abbrev(con)
        + ' '
        + rangeDesc(min, max)
        + ', but was '
        + ( con.called === 0
          ? 'not called.'
          : 'called ' + con.called + plural(con.called) + '.'
        )
        + '\ncontract defined at:'
      )
      err.type = 'contract'
      throw err
    }

   return {
    //soft validation, triggered before the
    id: 'isCalled',
    before: function (args) {
      //soft validation. fail if a call was
      //against the rule.
      if(max != null && this.called > max)
        error(this)
    },
    validate: function () {
      //hard check against the contract.
      //triggered manually, or on process.exit.
      if(!between(this.called, min, max))
        error(this)
    },
    update: function (_min, _max, change) {
      if(change) {
          min == null || (min += change)
          max == null || (max += change)
      } else
        min = _min; max = _max

      if(max != null && this.called > max)
        error(this)
    }
  }
}

exports.once = function () {
  return exports.isCalled.call(this, 1, 1)
}

exports.twice = function () {
  return exports.isCalled.call(this
